packages/playwright-ct-angular/registerSource.mjs

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-check
+// This file is injected into the registry as text, no dependencies are allowed.
+
+/** @typedef {import('../playwright-ct-core/types/component').ObjectComponent} ObjectComponent */
+
+import 'zone.js';
+import { reflectComponentType } from '@angular/core';
+import { getTestBed, TestBed } from '@angular/core/testing';
+import {
+  BrowserDynamicTestingModule,
+  platformBrowserDynamicTesting,
+} from '@angular/platform-browser-dynamic/testing';
+
+/** @type {WeakMap<import('@angular/core/testing').ComponentFixture, Record<string, import('rxjs').Subscription>>} */
+const __pwOutputSubscriptionRegistry = new WeakMap();
+
+/** @type {Map<string, import('@angular/core/testing').ComponentFixture>} */
+const __pwFixtureRegistry = new Map();
+
+getTestBed().initTestEnvironment(
+    BrowserDynamicTestingModule,
+    platformBrowserDynamicTesting(),
+);
+
+/**
+ * @param {ObjectComponent} component
+ */
+async function __pwRenderComponent(component) {
+  const componentMetadata = reflectComponentType(component.type);
+  if (!componentMetadata?.isStandalone)
+    throw new Error('Only standalone components are supported');
+
+  TestBed.configureTestingModule({
+    imports: [component.type],
+  });
+
+  await TestBed.compileComponents();
+
+  const fixture = TestBed.createComponent(component.type);
+  fixture.nativeElement.id = 'root';
+
+  __pwUpdateProps(fixture, component.props);
+  __pwUpdateEvents(fixture, component.on);
+
+  fixture.autoDetectChanges();
+
+  return fixture;
+}
```

Comment:
`TestBed.createComponent` discards the root element, which causes the locator to break. Suggested refactor below should resolve this issue:

```ts
/**
 * @param {ObjectComponent & MountOptions} component
 * @param {HTMLElement} rootElement
 */
async function __pwRenderComponent(component, rootElement) {
  await TestBed.compileComponents();

  const fixture = TestBed.createComponent(component.type)

  rootElement.replaceChildren(fixture.nativeElement)
  document.body.replaceChildren(rootElement)

  __pwUpdateProps(fixture, component.props);
  __pwUpdateEvents(fixture, component.on);

  fixture.autoDetectChanges();

  return fixture;
}
```

Reply:
Nice catch @chronospatian!
This is related to the behavior of Playwright's `#root >> internal:control=component` which will use `#root` if it has multiple children or the child if there is a single child.

We should indeed insert the element inside the root element provided by playwright.

The best way of handling this should be by implementing the `TestComponentRenderer` service:

```ts

class PlaywrightTestComponentRenderer extends TestComponentRenderer {

  constructor(rootElement) {
    super();
    this._rootElement = rootElement;
  }

  insertRootElement(testRootElementId) {
    const testRootElement = document.createElement('div');
    testRootElement.id = testRootElementId;
    this._rootElement.appendChild(testRootElement);
    return testRootElement;
  }

  removeRootElement(testRootElement) {
    this._rootElement.removeChild(testRootElement);
  }

}
...
TestBed.configureTestingModule({
  imports: [component.type],
  providers: [
    {
      provide: TestComponentRenderer,
      useValue: new PlaywrightTestComponentRenderer(rootElement)
    }
  ]
});

await TestBed.compileComponents();

const fixture = TestBed.createComponent(component.type);
...
```

Instead of setting the `root` id on the fixture.

Reply:
Here is the PR: https://github.com/sand4rt/playwright/pull/7
cc. @sand4rt